library/panic_unwind/src/wasm.rs

```diff
@@ -0,0 +1,32 @@
+//! Unwinding panics for wasm32.
+use alloc::boxed::Box;
+use core::any::Any;
+
+// The type of the exception payload that the wasm engine propagates
+// through unwinding for us. LLVM requires that it be a thin pointer.
+type Payload = Box<Box<dyn Any + Send>>;
+
+extern "C" {
+    /// LLVM lowers this intrinsic to the `throw` instruction.
+    #[link_name = "llvm.wasm.throw"]
+    fn wasm_throw(tag: i32, ptr: *mut u8) -> !;
+}
+
+pub unsafe fn panic(payload: Box<dyn Any + Send>) -> u32 {
+    // The payload we pass to `wasm_throw` will be exactly the argument we get
+    // in `cleanup` below. So we just box it up once, to get something pointer-sized.
+    let payload_box: Payload = Box::new(payload);
+    // The wasm `throw` instruction takes a "tag", which differentiates certain
+    // types of exceptions from others. LLVM currently just identifies these
+    // via integers, with 0 corresponding to C++ exceptions and 1 to C setjmp()/longjmp().
+    // Ideally, we'd be able to choose something unique for Rust, such that we
```

Comment:
What format does LLVM use for C++ exceptions on wasm? On native targets the exceptions have a "vendor" field, which rust fills in with `moz\0rust` to distinguish C++ and Rust exceptions. Maybe something similar can be done on wasm?

Reply:
Huhh. Yeah, that's true actually, this is a libunwind target. I guess it's up to whether we want to pull in libunwind on wasm or not; I guess it should be fine since it shouldn't be very much runtime stuff at all: https://github.com/llvm/llvm-project/blob/08eced5fccd2f103379292f119834a7a3c3b6b25/libunwind/src/Unwind-wasm.c

Reply:
Although, it seems like maybe llvm-libunwind only supports a single personality function per binary - since `__gxx_personality_wasm0` is hardcoded as the given personality to call.

Reply:
Ah, well, LLVM's codegen expects libunwind anyway. Alright.

Reply:
Which personality function is used with the PR in the current state? Where is it defined?

Does the format of `_Unwind_Exception` match that of other targets as used in `gcc.rs`? If so would it be possible to use the implementation in `gcc.rs`, but add a special case to call the `llvm.wasm.throw` llvm intrinsic instead of `_Unwind_RaiseException`? (or implement `_Unwind_RaiseException` using `llvm.wasm.throw` in the unwind crate?)

Reply:
> Which personality function is used with the PR in the current state? Where is it defined?

Right now, abort(), just like on msvc.

> would it be possible to use the implementation in `gcc.rs`, but add a special case to call the `llvm.wasm.throw` llvm intrinsic instead of `_Unwind_RaiseException`

Sorta; we'd basically still have to shim all of libunwind since llvm doesn't have a wasm32-unknown version of it. Maybe I'll try to keep this as-is for wasm32-unknown, but point wasi to the gcc version so it can be compatible with C++ code. (I don't think there's all that much C++ code being built for wasm32-unknown anyway; you basically need a custom sysroot for clang afaict). I think llvm only inserts `_Unwind_CallPersonality` in codegen for `wasm.get.selector()`, which we never generate, so that should be fine.

Reply:
panic_unwind only needs `_Unwind_RaiseException` and `_Unwind_DeleteException`. `_Unwind_RaiseException` is a call to `llvm.wasm.throw` and `_Unwind_DeleteException` a call to the `exception_cleanup` field of the exception. As such it should be trivial to implement both in way that will not cause issues if you were to link C++ code, unlike the current implementation.

> I don't think there's all that much C++ code being built for wasm32-unknown anyway; you basically need a custom sysroot for clang afaict

You can't right now because of an incompatibility with the C abi, but in the future once that is solved, it should be possible to link wasm32-unknown-unknown code with C and C++ code compiled by clang.

Reply:
> As such it should be trivial to implement both in way that will not cause issues if you were to link C++ code, unlike the current implementation.

It won't cause issues unless a C++ exception unwinds past a catch_unwind block (or a Rust panic unwinds past a C++ try/catch). In my eyes, that's fine for now? The C-unwind rfc says the former is undefined behavior, and the latter can be fixed pending llvm support.

Reply:
Though also, yeah, it would be pretty straightforward to just implement it in terms of _Unwind_Exception.